Docs 13105 local runtime registry (#586)
mulesoft/docs-runtime-fabric: modules/ROOT/pages/configure-local-registry.adoc

= Using a Local Registry with Runtime Fabric on Self-Managed Kubernetes

Runtime Fabric uses its own registry to store the necessary Docker images for installing and managing Runtime Fabric and for running applications in Kubernetes. If your security policies require that you pull your Docker images from a registry under your control, you can set up a local registry to pull and store these Docker images.

== Limitations on Local Registry Support 

Review the following limitations on support for local registries: 

> * Runtime Fabric supports local registries on Self-Managed Kubernetes only.

Reviewer: ```suggestion
* Runtime Fabric supports local registries on self-managed Kubernetes only.
```

Author: See slack note

+
---- 
# docker login $RTF_DOCKER_SERVER --<username> $RTF_DOCKER_USERNAME --<password> $RTF_DOCKER_PASSWORD → docker pull rtf-runtime-registry.kqa.msap.io/mulesoft/rtf-agent:v1.12.0 → docker tag rtf-runtime-registry.kqa.msap.io/mulesoft/rtf-agent:v1.12.0 localhost:5000/mulesoft/rtf-agent:v1.12.0 → docker push localhost:5000/mulesoft/rtf-agent:v1.12.0
----

. Finalize the Runtime Fabric installation, and configure it to pull from your local registry:
+
----
# rtfctl install ‘<activation_data>’ --image-registry localhost:5000 
----
+
Alternatively, if you use authentication to access your registry, use the following command:

----
# rtfctl install ‘<activation_data>’ --image-registry localhost:5000 --image-pull-secret <your-local-reg-secret>
----

This command ensures that Runtime Fabric creates the necessary overrides for the local registry URL and `image-pull-secret` values so Runtime Fabric continues to pull images from the local registry even after an upgrade.

The `rtfctl uninstall` command deletes any pull secret you specify during installation.

When using `rtfctl` to perform a backup and restore procedure, the backups point to your local registry URL and pull secret.

> * When using `rtfctl` to perform a backup and restore procedure, the backups point to your local registry URL and pull secret. 

Reviewer: duplicated

> * When using `rtfctl` to perform a backup and restore procedure, the backups point to your local registry URL and pull secret. 

Author: ```suggestion
```

> = Using A Local Registry with Runtime Fabric on Self-Managed Kubernetes

Reviewer: ```suggestion
= Using a Local Registry with Runtime Fabric on Self-Managed Kubernetes
```

> Runtime Fabric uses its own registry to store the necessary Docker images for installing and managing Runtime Fabric and for running applications in Kubernetes. If your security policies require that you pull your Docker images from a registry under your control, you can set up a local registry to pull and store these Docker images.

Reviewer: Just a suggestion for first use on a page: Anypoint Runtime Fabric.

> == Limitations for Local Registry Support 

Reviewer: I actually had to look this up, because I want to say "limitations to":
```suggestion
== Limitations on Local Registry Support 
```

> Review the following limitations about support for local registries: 

Reviewer: ```suggestion
Review the following limitations on support for local registries: 
```

> * Runtime Fabric supports local registries with Runtime Fabric on Self-Managed Kubernetes only.

Reviewer: Does this mean the same thing:
```suggestion
* Runtime Fabric supports local registries on Self-Managed Kubernetes only.
```

> * You can use a local registry only if you create a new Runtime Fabric using version 1.12.0 or higher. You cannot add a local registry to an existing Runtime Fabric installation (regardless of its version) even if you upgrade that installation to 1.12.0 or higher. 

Reviewer: ```suggestion
* You can use a local registry only if you create a new Runtime Fabric instance of version 1.12.0 or later. You cannot add a local registry to an existing Runtime Fabric installation (regardless of its version) even if you upgrade that installation to 1.12.0 or later. 
```

> * You must use xref:install-rtfctl.adoc[Runtime Fabric command line utility (`rtfctl`)] version 0.3.150 or higher.

Reviewer: ```suggestion
* You must use xref:install-rtfctl.adoc[Runtime Fabric command line utility (`rtfctl`)] version 0.3.150 or later.
```

Reviewer: I have both RF "installation" and "instance" rolling around in my head, depending on the syntax of the sentence. You choose. Just be sure to replace all instances of "the" or "a" Runtime Fabric (name) and plural "Fabrics," okay?

> * If you create a Runtime Fabric to use with a local registry, you cannot later reconfigure it to use the `rtf-runtime-registry` endpoint. To use the `rtf-runtime-registry` endpoint, you must install a new Runtime Fabric using the standard installation procedure.

Reviewer: a Runtime Fabric installation? instance?

Reviewer: Ditto for "a new...."

* You're responsible for synchronization between your registry and the Runtime Fabric registry. MuleSoft doesn't provide additional software to synchronize images between the registries.

> * If you use a local registry that needs authentication, synchronization and propagation of a pull-secret between different namespaces is your responsibility. Alternatively, Runtime Fabric can synchronize your pull-secret across different namespaces if your secret uses the following label:

Reviewer: I looked this up: "pull secret" is not hyphenated as a noun.

----
  labels:
    rtf.mulesoft.com/synchronized: "true"
----

> * If you need to authenticate access to your local registry, ensure the https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#registry-secret-existing-credentials[corresponding credentials are configured as a Kubernetes secret in the `rtf-` namespace^]. You'll need the name of the configured secret when you install Runtime Fabric. 

Reviewer: ensure > ensure that

* You also need to synchronize in your registry the Mule runtime versions you intend to use for application deployments. Refer to the xref:release-notes::runtime-fabric/runtime-fabric-runtimes-release-notes.adoc[Runtime Fabric Patch Update Release Notes] for information about the latest Mule runtime patch versions. 

== Configure a Local Registry for Use with Runtime Fabric 

> To configure a local registry you must gather and add the necessary credentials to synchronize the registries.

Reviewer: ```suggestion
To configure a local registry, you must gather and add the necessary credentials to synchronize the registries.
```

> === Before You Begin

Reviewer: Making "Before You Begin" a subtopic of the task causes an information hierarchy issue if the text following the prerequisites does not have a suitable heading:

= Using...
== Limitations
== Configure 
=== Before You Begin
[=== steps title? Steps should not be part of Before You Begin....]
== See Also

Reviewer: One thing we could do...that I suggest we DO do...is make the prereqs. part of the intro to the steps: 

To configure a local registry, you must gather and add the necessary credentials to synchronize the registries. However, before you do that, ensure that you've already performed the following tasks:

(Are we assuming that the audience knows how to do these things? Do we need to link to them?)

> Ensure you've performed the following tasks: 

Reviewer: See suggestion above and note "ensure _that_...."

> . Setup, configure, and test a local Docker image registry.

Reviewer: Verb form is "to set up":
```suggestion
. Set up, configure, and test a local Docker image registry.
```

. Synchronize to your local registry all the Docker images you need to install Runtime Fabric:
+
[NOTE]

> Dependency versions are specific to the Runtime Fabric version.

Reviewer: I suggest moving this so that it does not interrupt the flow between the list introduction and the list items.

 
+
* appInit: `mulesoft/rtf-app-init:v1.0.35`
* muleClusterIpService: `mulesoft/rtf-mule-clusterip-service:v1.2.36`
* resourceFetcher: `mulesoft/rtf-resource-fetcher:v1.0.46`
* nginx: `mulesoft/base-image-nginx-1.21.1:v1.1.9`
* monitoringSidecar: `mulesoft/dias-anypoint-monitoring-sidecar:v1.3.18`
* clusterOps: `mulesoft/rtf-cluster-ops:v1.1.25`
* securityFabricEdge: `mulesoft/securityfabric-edge:v1.1.346`
* externalLogForwarder: `mulesoft/rtf-pkg-fluentbit:v1.2.53`
* coreAction: `mulesoft/rtf-core-actions:v1.0.11`
* persistenceGateway: `mulesoft/rtf-object-store:v1.0.44`
* rtfDaemon: `mulesoft/rtf-daemon:v1.0.3`

> 

Reviewer: This is where we need to transition from background info to the actual task, which gets lost being included with the Before You Begin narrative.

. xref:install-self-managed.adoc#step-3-download-the-rtfctl-utility[Download and install the `rtfctl` command line utility] version 0.3.150 or higher.

> . xref:install-self-managed.adoc#step-3-download-the-rtfctl-utility[Create a Runtime Fabric using Runtime Manager], and gather the activation data. 

Reviewer: ditto re. "a Runtime Fabric"

----

. To pull and synchronize images, run the following command, replacing the `<username>` and `<password>` values where appropriate:
+
---- 
# docker login $RTF_DOCKER_SERVER --<username> $RTF_DOCKER_USERNAME --<password> $RTF_DOCKER_PASSWORD → docker pull rtf-runtime-registry.kqa.msap.io/mulesoft/rtf-agent:v1.12.0 → docker tag rtf-runtime-registry.kqa.msap.io/mulesoft/rtf-agent:v1.12.0 localhost:5000/mulesoft/rtf-agent:v1.12.0 → docker push localhost:5000/mulesoft/rtf-agent:v1.12.0
----

. Finalize the Runtime Fabric installation, and configure it to pull from your local registry:
+
----
# rtfctl install ‘<activation_data>’ --image-registry localhost:5000 
----
+
Alternatively, if you use authentication to access your registry, use the following command:

----
# rtfctl install ‘<activation_data>’ --image-registry localhost:5000 --image-pull-secret <your-local-reg-secret>
----

This command ensures that Runtime Fabric creates the necessary overrides for the local registry URL and `image-pull-secret` values so Runtime Fabric continues to pull images from the local registry even after an upgrade.

> The `rtfctl uninstall` command deletes any pull-secret you specify during installation.

Reviewer: ```suggestion
The `rtfctl uninstall` command deletes any pull secret you specify during installation.
```

> When using `rtfctl` to perform a back up and restore procedure, the backups will point to your local registry URL and pull secret.

Reviewer: "backup" is one word as an adjective
```suggestion
When using `rtfctl` to perform a backup and restore procedure, the backups point to your local registry URL and pull secret.
```

> * When using `rtfctl` to perform a back up and restore procedure, the backups will point to your local registry url and pull secret. 

Reviewer: ```suggestion
* When using `rtfctl` to perform a backup and restore procedure, the backups point to your local registry URL and pull secret. 
```

+
[NOTE]
Dependency versions are specific to the Runtime Fabric version.
 

. xref:install-self-managed.adoc#step-3-download-the-rtfctl-utility[Download and install the `rtfctl` command line utility] version 0.3.150 or higher.
. xref:install-self-managed.adoc#step-3-download-the-rtfctl-utility[Create a Runtime Fabric using Runtime Manager], and gather the activation data. 

=== Configure Your Local Registry

. Obtain `rtf-runtime-registry` credentials:

----
# rtfctl get docker-registry-login <activation-data>
----

This command sets `RTF_DOCKER_SERVER`, `RTF_DOCKER_USERNAME`, and `RTF_DOCKER_PASSWORD` in the current shell environment. 

. Run the following commands to verify that the Docker login to the `rtf-runtime-registry` succeeds:
+
---- 
# eval $(rtfctl get docker-registry-login)
# docker login $RTF_DOCKER_SERVER --<username> $RTF_DOCKER_USERNAME --<password> $RTF_DOCKER_PASSWORD
----

. To pull and synchronize images, run the following command, replacing the `<username>` and `<password>` values where appropriate:
+
---- 

> # docker login $RTF_DOCKER_SERVER --<username> $RTF_DOCKER_USERNAME --<password> $RTF_DOCKER_PASSWORD → docker pull rtf-runtime-registry.kqa.msap.io/mulesoft/rtf-agent:v1.12.0 → docker tag rtf-runtime-registry.kqa.msap.io/mulesoft/rtf-agent:v1.12.0 localhost:5000/mulesoft/rtf-agent:v1.12.0 → docker push localhost:5000/mulesoft/rtf-agent:v1.12.0

Reviewer 2: Are we going to change this to prod URLs?

Author: @rohit-verma7 Good catch, we should change this to whatever is accurate
